vsphere/host_data_store_system_helper.go

```diff
@@ -0,0 +1,147 @@
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	return hs.ConfigManager().DatastoreSystem(ctx)
+}
+
+// availableScsiDisk checks to make sure that a disk is available for use in a
+// VMFS datastore, and returns the ScsiDisk.
+func availableScsiDisk(dss *object.HostDatastoreSystem, name string) (*types.HostScsiDisk, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	disks, err := dss.QueryAvailableDisksForVmfs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot query available disks: %s", err)
+	}
+
+	var disk *types.HostScsiDisk
+	for _, d := range disks {
+		if d.CanonicalName == name {
+			disk = &d
+			break
+		}
+	}
+	if disk == nil {
+		return nil, fmt.Errorf("%s does not seem to be a disk available for VMFS", name)
+	}
+	return disk, nil
+}
+
+// diskSpecForCreate checks to make sure that a disk is available to be used to
+// create a VMFS datastore, specifically in its entirety, and returns a
+// respective VmfsDatastoreCreateSpec.
+func diskSpecForCreate(dss *object.HostDatastoreSystem, name string) (*types.VmfsDatastoreCreateSpec, error) {
+	disk, err := availableScsiDisk(dss, name)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	options, err := dss.QueryVmfsDatastoreCreateOptions(ctx, disk.DevicePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not get disk creation options for %s: %s", name, err)
+	}
+	var option *types.VmfsDatastoreOption
+	for _, o := range options {
+		if _, ok := o.Info.(*types.VmfsDatastoreAllExtentOption); ok {
+			option = &o
+			break
+		}
+	}
+	if option == nil {
+		return nil, fmt.Errorf("cannot use entire disk on device %s for datastore", name)
+	}
+	return option.Spec.(*types.VmfsDatastoreCreateSpec), nil
+}
+
+// diskSpecForExtend checks to make sure that a disk is available to be
+// used to extend a VMFS datastore, specifically in its entirety, and returns a
+// respective VmfsDatastoreExtendSpec if it is. An error is returned if it's
+// not.
+func diskSpecForExtend(dss *object.HostDatastoreSystem, ds *object.Datastore, name string) (*types.VmfsDatastoreExtendSpec, error) {
+	disk, err := availableScsiDisk(dss, name)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	options, err := queryVmfsDatastoreExtendOptions(ctx, dss, ds, disk.DevicePath, true)
+	if err != nil {
+		return nil, fmt.Errorf("could not get disk extension options for %s: %s", name, err)
+	}
+	var option *types.VmfsDatastoreOption
+	for _, o := range options {
+		if _, ok := o.Info.(*types.VmfsDatastoreAllExtentOption); ok {
+			option = &o
+			break
+		}
+	}
+	if option == nil {
+		return nil, fmt.Errorf("cannot use entire disk on device %s for datastore", name)
```

Comment:
can we switch out the formatting argument to `%q` and provide a better reason for this?